pymatviz/ptable/ptable_matplotlib.py

```diff
@@ -1,888 +1,407 @@
+    PTableProjector,
 )
-from pymatviz.utils import df_ptable, pick_bw_for_contrast, si_fmt
+from pymatviz.utils import get_cbar_label_formatter, pick_bw_for_contrast
 
 
 if TYPE_CHECKING:
     from typing import Any, Callable
 
+    import matplotlib.pyplot as plt
+    from matplotlib.colors import Colormap
     from matplotlib.typing import ColorType
+    from pymatgen.core import Element
 
-    from pymatviz.utils import ElemValues
+    from pymatviz.ptable._process_data import PTableData
 
+# Custom types
+ElemStr = str  # element as a str
+ElemValues = Union[dict[Union[str, int], float], pd.Series, Sequence[str]]
 
-class PTableProjector:
-    """Project (nest) a custom plot into a periodic table.
 
-    Attributes:
-        cmap (Colormap | None): Colormap.
-        data (pd.DataFrame): Data for plotting.
-        norm (Normalize): Data min-max normalizer.
-        hide_f_block (bool): Whether to hide f-block.
-        elem_types (set[str]): Types of elements present.
-        elem_type_colors (dict[str, str]): Element typed based colors.
-        elem_colors (dict): Element based colors.
+def ptable_heatmap(
+    data: pd.DataFrame | pd.Series | dict[str, list[list[float]]] | PTableData,
+    *,
+    # Heatmap specific
+    colormap: str = "viridis",
+    exclude_elements: Sequence[str] = (),
+    overwrite_tiles: dict[ElemStr, OverwriteTileValueColor] | None = None,
+    infty_color: ColorType = "lightskyblue",
+    nan_color: ColorType = "lightgrey",
+    log: bool = False,
+    sci_notation: bool = False,
+    tile_size: tuple[float, float] = (0.75, 0.75),  # TODO: WIP, don't use
+    # Figure-scope
+    on_empty: Literal["hide", "show"] = "show",
+    hide_f_block: bool | Literal["AUTO"] = "AUTO",
+    f_block_voffset: float = 0,  # TODO: WIP, don't use
+    plot_kwargs: dict[str, Any] | None = None,
+    # Axis-scope
+    ax_kwargs: dict[str, Any] | None = None,
+    text_colors: Literal["AUTO"] | ColorType | dict[ElemStr, ColorType] = "AUTO",
+    # Symbol
+    symbol_pos: tuple[float, float] | None = None,
+    symbol_kwargs: dict[str, Any] | None = None,
+    # Value
+    value_show_mode: Literal["value", "fraction", "percent", "off"] = "value",
+    value_pos: tuple[float, float] | None = None,
+    value_fmt: str = "AUTO",
+    value_kwargs: dict[str, Any] | None = None,
+    # Colorbar
+    show_cbar: bool = True,
+    cbar_coords: tuple[float, float, float, float] = (0.18, 0.8, 0.42, 0.05),
+    cbar_range: tuple[float | None, float | None] = (None, None),
+    cbar_label_fmt: str = "AUTO",
+    cbar_title: str = "Element Count",
+    cbar_title_kwargs: dict[str, Any] | None = None,
+    cbar_kwargs: dict[str, Any] | None = None,
+    # Deprecated args, don't use
+    colorscale: str | None = None,
+    heat_mode: Literal["value", "fraction", "percent"] | None = None,
+    show_values: bool | None = None,
+    fmt: str | None = None,
+    cbar_fmt: str | None = None,
+    show_scale: bool | None = None,
+) -> plt.Figure:
```

Comment:
i think returning a `plt.Figure` instead of `plt.Axes` is a big breaking change. not something to do lightly. i realize it makes the `matplotlib` `ptable` plotters more consistent, but a lot of the `pymatviz` code i and others like @CompRhys wrote relies on `ptable_heatmap` returning `Axes`. at the least, i think we would need to try to make an automatic migration.

Reply:
Okay! Sure. I would revert this change (as current projector implementation would return the `Figure`), I would have a look and try to make it backwards compatible

Reply:
I just added a `return_type` (default to `axes` to be backwards compatible) in 0f9bdc796f8f90b6c92d344e062a4d2479d75117 to encourage the migration from `axes` to `figure` within one year period, hopefully this could make this transition smoother. 

I have to note that because of the change in `axes` order, the resulted figure could be suboptimal though:
```python

from matminer.datasets import load_dataset

from pymatviz.enums import Key
from pymatviz.io import save_and_compress_svg
from pymatviz.process_data import count_elements
from pymatviz.ptable import ptable_heatmap


df_expt_gap = load_dataset("matbench_expt_gap")


# Elemental Plots
ax = ptable_heatmap(count_elements(df_expt_gap[Key.composition]), log=True)
title = (
    f"Elements in Matbench Experimental Band Gap ({len(df_expt_gap):,} compositions)"
)
ax.set_title(title, y=0.96, fontsize=16, fontweight="bold")
save_and_compress_svg(ax, "ptable-heatmap")
```

Where the title kind of overlap with the body of the figure (changing the `y` doesn't seem to move the title, I guess it's owing to the position of the ax).

![ptable-heatmap](https://github.com/janosh/pymatviz/assets/80093591/105255bb-0540-4e13-9200-fd69e758b684)

Reply:
i just tried it. changing `y` in `ax.set_title()` still has an effect but it's much less pronounced than before as the units are relative to the size of the axes (which before was the whole figure and now is just one element tile)

Reply:
Yes exactly, it seems the new implementation changed the "scope" of the axes. Sorry it looks like a wrong timing to propose this breaking change (though the `return_type` arg should make it non-breaking, but less pretty).

In current implementation, the `ax` is generated via: https://github.com/janosh/pymatviz/blob/2c4c5da15ea1ffe6cb85f1092b0ab8519c73eb94/pymatviz/ptable/ptable_matplotlib.py#L737-L739

With the new implementation, we now switch to `plt.subplots` to generate the `figure` and `axes`:
```python
self.fig, self.axes = plt.subplots(n_periods, n_groups, **plot_kwargs)
```

Upon a closer look, I was able to "fix" the inconsistency by using a huge relative position:
```python
from matminer.datasets import load_dataset

from pymatviz.enums import Key
from pymatviz.io import save_and_compress_svg
from pymatviz.process_data import count_elements
from pymatviz.ptable import ptable_heatmap


df_expt_gap = load_dataset("matbench_expt_gap")


# Elemental Plots
ax = ptable_heatmap(count_elements(df_expt_gap[Key.composition]), log=True)
title = (
    f"Elements in Matbench Experimental Band Gap ({len(df_expt_gap):,} compositions)"
)
ax.set_title(title, x=0.75, y=2.5, fontsize=18, fontweight="bold")
save_and_compress_svg(ax, "ptable-heatmap")
```
![debug](https://github.com/user-attachments/assets/2639481e-d6f8-4a65-8100-3240278b7730)

So I would say the new implementation is still "compatible" with previous version (at the price of having to tweak the title position), but I wish to encourage users to migrate to `figure` over the one year period.

Do you have further comment or can you review (and possibly merge) this? Thanks!